### PR TITLE
Smarter rate limiter using partial amounts

### PR DESCRIPTION
**Pull request by sunbreak1211**

### CODE REVIEW

File: src/funnels/Swapper.sol
Comment (oldchili) [inline src/funnels/Swapper.sol]: Don't think we need to update zzz on each operation.

Comment (sunbreak1211) [reply src/funnels/Swapper.sol]: The idea to put it there is to make sure both will be saved in the same SSTORE (as they share the same slot) so it is just easier to do the assignment always.

Comment (telome) [reply src/funnels/Swapper.sol]: Are you sure they share the same slot? I would have thought that the first (uint64, uint128, uint64) are packed into one uint256 slot and then the last uint128 is packed into a second uint256 slot.

Comment (sunbreak1211) [reply src/funnels/Swapper.sol]: Yes you are right, I need to reorder that.

Comment (sunbreak1211) [reply src/funnels/Swapper.sol]: Done.

Comment (oldchili) [reply src/funnels/Swapper.sol]: It might be nicer to assign to zzz and amt in the same order in both unchecked blocks.
Also - did you check to see if it is is indeed done as one write operation here?

Comment (sunbreak1211) [reply src/funnels/Swapper.sol]: Order changed. Anyway the slots (and setting of variables) order is one of the main things I want to review for both Swapper and Depositor and their automation contracts.
Regarding the one write operation, I didn't check it for this code but I had done in the past. I'm 99% sure with at least the optimizer it should just write once but I'll double check with this code.

File: src/funnels/Depositor.sol
Comment (telome) [inline src/funnels/Depositor.sol]: I'm wondering if she should maybe reset the `limits[gem0][gem1].amt{i}` here when `cap{i}` is changed so that a reduction of `cap{i}` cannot temporarily be ignored in case  the existing `amt{i} > newCap{i}`. 

Or alternatively I guess we could just leave a note warning that a change of cap will only become effective after (up to) `hop` amount of time.

Comment (sunbreak1211) [reply src/funnels/Depositor.sol]: Changed it in both Depositor and Swapper, we might evaluate to directly set the amounts to 0 or just make both values as arguments and let the subdao proxy to define the values depending on the needs of the case.

Comment (oldchili) [reply src/funnels/Depositor.sol]: I think current code in the Depositor could bring a weird state where one amt is updated to the cap and one is not, maybe just simpler to indeed always reset to cap (also in the Swapper for alignment).

File: src/funnels/Swapper.sol
Comment (oldchili) [inline src/funnels/Swapper.sol]: The comment is wrong as it is the amount remaining.
Maybe rename to something like `left` everywhere? (to remove also remove confusion with the StableDepositor's amts, which are something different).

File: src/funnels/Swapper.sol
Comment (oldchili) [inline src/funnels/Swapper.sol]: Not a huge save but note that moving memory to stack  saves gas for variables that are read multiple times or written to temporarily.

In this case you can consider using something like:
```
uint64 _zzz = limit.zzz;
uint128 _amt = limit.amt;
```

File: src/funnels/Swapper.sol
Comment (oldchili) [inline src/funnels/Swapper.sol]: I know we said these 2 lines are grouped to one storage op, but for some reason I see that when commenting out the 2nd line we get a 20K gas saving (180359 for just the first swap in testSwap).

Also note that the base code takes 181131 gas for the first swap in testSwap, while the current one (without that line elimination) takes 206298.
I wouldn't expect such a big diff as I don't think there should be more storage write ops.

File: src/funnels/automation/StableSwapper.sol
Comment (oldchili) [inline src/funnels/automation/StableSwapper.sol]: Don't forget to add documentation to the new vars (can be done at the end).

File: src/funnels/Swapper.sol
Comment (sunbreak1211) [reply src/funnels/Swapper.sol]: I think this is understood now. Commenting the `zzz` makes a change because in this specific test is the only variable actually changing. So at least the first part should be solved now, not sure if I'm following the second paragraph.

File: src/funnels/Swapper.sol
Comment (sunbreak1211) [reply src/funnels/Swapper.sol]: Yeah not sure about this is worth it, I tried setting all original to stack with variables instead of initiating to the memory struct and that is more expensive. So I guess it some kind of combination what makes it the most efficient. I'd vote to leave it as it is.

Comment (oldchili) [reply src/funnels/Swapper.sol]: ok

File: src/funnels/Swapper.sol
Comment (sunbreak1211) [reply src/funnels/Swapper.sol]: For the second one, after discussing I realized it has to be due to a storage warming happening on `setLimits` called by the `setUp` in the old structure, as the only slot gets written. However in the new structure the slot having the `amt/zzz` shouldn't be written in `setLimits` as `amt <= cap`.

File: src/funnels/Swapper.sol
Comment (sunbreak1211) [reply src/funnels/Swapper.sol]: Fixed.

Comment (sunbreak1211) [reply src/funnels/Swapper.sol]: And changed variable to `due`.

File: src/funnels/automation/StableSwapper.sol
Comment (sunbreak1211) [reply src/funnels/automation/StableSwapper.sol]: Done!

File: src/funnels/Depositor.sol
Comment (sunbreak1211) [reply src/funnels/Depositor.sol]: Hmmm just wondering if that would be a problem at all.
The due amounts do not enforce any type of relationship between them on deposit/withdraw, then not sure it would be a problem when one gets updated to its `cap` and the other keeps unchanged. However I agree that might be easier to reason to just set both to 0.

Comment (oldchili) [reply src/funnels/Depositor.sol]: when you say 0 you mean set both to the new `cap`s right?

Comment (sunbreak1211) [reply src/funnels/Depositor.sol]: Yeah I guess both are valid options, we can set to 0 and wait for the new `hop` to pass (if didn't) or set to `cap`, which might allow two operations for the full `cap` pretty soon if the time to surpass the new `hop` is close to happen.
That's was the main reason why I felt keeping the `due` amount or setting to new `cap` if is greater was the most fair situation.

Comment (telome) [reply src/funnels/Depositor.sol]: To keep things simple, I'd be tempted to reset the state in a way that looks similar to the initial state: set the `due` amounts to 0 and set `zzz` to 0 as well. 

That means that it will be possible to spend `prevDue` + `newCap` back to back but that seems like an acceptable trade-off occurring at the boundaries between new configs.

Comment (sunbreak1211) [reply src/funnels/Depositor.sol]: Done.

File: src/funnels/Depositor.sol
Comment (telome) [inline src/funnels/Depositor.sol]:
```suggestion
    mapping (address => mapping (address => mapping (uint24 => PairLimit))) public limits; // Rate limit parameters per (gem0, gem1, fee) pool
```

File: src/funnels/Depositor.sol
Comment (telome) [inline src/funnels/Depositor.sol:125]: Should the limit really be set for a fee level? Maybe we could leave the choice of the fee level to the facilitator?

File: src/funnels/automation/StableDepositor.sol
Comment (telome) [inline src/funnels/automation/StableDepositor.sol]: One open question is whether it could make sense to make this a signed integer representing the amount of times `deposit()` can be called when positive, or the amount of time `withdraw()` can be called when negative. Right now we decrement the same `num` for both deposits and withdrawals but maybe we should be more consistent with how we treat the `num` in the StableSwapper where we have a different `num` for each swap direction (and generally expect one of the two swap directions to have its `num == 0`).

File: src/funnels/Depositor.sol
Comment (sunbreak1211) [reply src/funnels/Depositor.sol:125]: Hmmm this is what we discussed in chat, I think both options are valid. I tend more to feel choosing the pool is something that is not needed to be decided by the facilitator as it will split liquidity and probably requires more strategical level of decision than a tactical one.

File: src/funnels/automation/StableDepositor.sol
Comment (sunbreak1211) [reply src/funnels/automation/StableDepositor.sol]: This one is interesting, I think it actually makes a lot of sense the facilitator enforces the direction to go. Being the keeper the one that chooses it seems misleading.

File: src/funnels/Depositor.sol
Comment (sunbreak1211) [reply src/funnels/Depositor.sol]: Fixed!

File: src/funnels/Depositor.sol
Comment (telome) [reply src/funnels/Depositor.sol:125]: Yes it's not so much because I think the facilitator should decide the value of that param, but more that it may be useful for it to be changeable rapidly by the SubDAO, without waiting for an executive, and have the facilitator just be enacting the will of the SubDAO. In general I guess anything that is not rate-limit-related could be made easy to update without an executive unless there is a good reason not to.

Comment (sunbreak1211) [reply src/funnels/Depositor.sol:125]: Yeah I think there are good arguments for both ways, willing to change if preferred. Let's see what @oldchili thinks about this.

Comment (oldchili) [reply src/funnels/Depositor.sol:125]: Yeah I don't have a strong opinion.
Maybe slightly leaning to keep this change just to better support the case of a slow migration from one pool to another.
In that case if the rate limiting is combined it becomes a bit harder to manage (need the facilitator to only use half of `due` for each pool, etc etc..).

Note that if the subdao wants the freedom of deciding which pool out of two to use without an exec it can set a limit for both pools in advance. The downsize for that is that it doubles the risk that the rate limit is there to protect against, but I think that's something manageable.

File: src/funnels/automation/StableDepositor.sol
Comment (oldchili) [reply src/funnels/automation/StableDepositor.sol]: Yeah makes sense.
Think of a keeper script that just simulates a tx and calls it on chain in case it doesn't revert.
How would it know what function to call (deposit or withdraw) without this?

I personally don't like much using signed integers, so would prefer another variable to indicate the direction (especially as it seem we have some place in the packed struct).
Can check it in the same require as we check `cfg.num`.

Comment (sunbreak1211) [reply src/funnels/automation/StableDepositor.sol]: The good thing about the signed integer is that easily enforces that they are mutually exclusive which I'd think it makes more sense. But if you insist in two variables I can give a try, worst case we can split in two uint16 which allows much less number but it is still more than enough for this use case.
Anyway I think I'd prefer to manage this change in a specific PR once this one gets merged.

File: src/funnels/Swapper.sol
Comment (oldchili) [inline src/funnels/Swapper.sol]: What do you think about changing `hop` to `era` in Swapper and Depositor? `hop` is usually an interval between operations, while `era` is a time interval where several events can happen.
I'd also rename `zzz`, as that indicates a sleeping time, maybe to `beg` to indicate the beginning?
With these renamings we also void confusion with the automation contracts, which have `hop` and `zzz`, but with the intended meanin.

Comment (oldchili) [reply src/funnels/Swapper.sol]: We also need to update the documentation of the variables, as it still just assumes one swap in the beginning of the `hop` period everywhere.

File: src/funnels/Swapper.sol
Comment (oldchili) [inline src/funnels/Swapper.sol]:
```suggestion
    event Swap(address indexed sender, address indexed src, address indexed dst, uint256 amt, uint256 out);
```

File: src/funnels/Depositor.sol
Comment (oldchili) [inline src/funnels/Depositor.sol]:
```suggestion
    mapping (address => uint256) public wards;                                             // Admins
```
align spacing

File: src/funnels/automation/StableSwapper.sol
Comment (oldchili) [inline src/funnels/automation/StableSwapper.sol]: This was named `req` (required) on purpose to differentiate it from the keeper's `minOut` (same for StableDepositor).

File: src/funnels/automation/StableDepositor.sol
Comment (oldchili) [inline src/funnels/automation/StableDepositor.sol]:
```suggestion
            liquidity  : 0,         // Use desired amounts
```
A bit of better spacing

File: src/funnels/automation/StableDepositor.sol
Comment (oldchili) [inline src/funnels/automation/StableDepositor.sol]:
```suggestion
            liquidity  : 0,         // Use desired amounts
```

File: src/funnels/Depositor.sol
Comment (telome) [reply src/funnels/Depositor.sol:125]: Ok let's keep it as-is. Resolving.

File: src/funnels/Swapper.sol
Comment (telome) [reply src/funnels/Swapper.sol]: +1 to using `era` and whatever else for the running timestamp (`beg` is fine).

File: src/funnels/automation/StableSwapper.sol
Comment (sunbreak1211) [reply src/funnels/automation/StableSwapper.sol]: Changed.

File: src/funnels/Swapper.sol
Comment (sunbreak1211) [reply src/funnels/Swapper.sol]: Changed, I used `end` changing a bit the logic (which I think is more efficient).

File: src/funnels/automation/StableDepositor.sol
Comment (sunbreak1211) [reply src/funnels/automation/StableDepositor.sol]: Ended up making this sub PR to this PR: https://github.com/makerdao/dss-allocator/pull/29

Comment (sunbreak1211) [reply src/funnels/automation/StableDepositor.sol]: Let's see if we like it or we prefer to explore the other option

File: src/funnels/Depositor.sol
Comment (sunbreak1211) [reply src/funnels/Depositor.sol:233]: I'm actually using the order of how I'm reading from the structure.

Comment (sunbreak1211) [reply src/funnels/Depositor.sol:233]: First the `cap`s, then the `era`.

**Review by oldchili — APPROVED** [review]
LGTM

File: src/funnels/Depositor.sol
Comment (telome) [inline src/funnels/Depositor.sol]:
```suggestion
        uint32  end; // Timestamp of when the current batch ends for a (gem0, gem1, fee) pool
```

File: src/funnels/Swapper.sol
Comment (telome) [inline src/funnels/Swapper.sol]:
```suggestion
        uint32  end; // Timestamp of when the current batch ends
```

**Review by telome — APPROVED** [review]
Looks good

**Pull request merged**
